docs/proposals/004-initialization.md

```diff
@@ -0,0 +1,191 @@
+  or if the member has an initialization expression that defines its default value.
+- It is a sized-array type where the element type is default-initializable.
+- It is a tuple type where all element types are default-initializable.
+- It is an `Optional<T>` type for any `T`.
+
+This means that an unsized-array type or an existential type, or any composite type that contains such types without providing an explicit default constructor are not considered default-initializable.
+
+Note that `void` type should be treated as default initializable, since `void` type can have only one value that is `()` (empty tuple).
+
+### Variable Initialization
+
+If the type of a local variable is default-initializable, then its default initializer will be invoked at its declaration site implicitly to intialize its value:
+```c++
+int x; // x will be default initialized to 0 because `int` is default-initializable.
+// The above is equivalent to:
+int x = int();
+
+struct S { int x; int y; }
+S s; // s will be default initialized to {0, 0} because `S` is default-initializable.
+```
+
+If a type is not default-initializable, and the declaration site does not provide an intial value for the variable, the compiler should generate an error:
+```csharp
+struct V { int[] arr; }
+
+V v; // error: `v` must be initialized at declaration.
+```
+
+For backward compatibility, we will introduce a compiler option to turn this error into a warning, but we may deprecate this option in the future.
+
+### Generic Type Parameter
+
+A generic type parameter is not considered default-initializable by-default. As a result, the following code should produce error:
+```csharp
+void foo<T>()
+{
+    T t; // error, `t` is uninitialized.
+}
+```
+
+### Automatic Synthesis of Default-Initializer
+
+If a `struct` type is determined to be default-initializable but a default constructor isn't explicitly provided by the user, the Slang compiler should
+synthesize such a constructor for the type. The synthesis logic should be recursively invoke default initializer on all members.
+
+### Automatic Synthesis of `IDefaultInitializable` Conformance
+
+If a `struct` type provides an explicit default constructor, the compiler should automatically add `IDefaultIinitializable` to the conformance list of
+the type, so it can be used for any generic parameters constrained on `IDefaultInitializble`.
+
+### Initialization List
+
+Slang allows initialization of a variable by assigning it with an initialization list. 
+Generally, Slang will always try to resolve initialization list coercion as if it is an explicit constructor invocation.
+For example, given:
+```
+S obj = {1,2};
+```
+Slang will try to convert the code into:
+```
+S obj = S(1,2);
+```
+
+As a special case, an empty initializer list will translate into a default-initialization:
+```csharp
+S obj = {};
+// equivalent to:
+S obj = S();
```

Comment:
I just want to be very clear that

`S obj = {}; `

won't zero-initialize any members of S. It's not a C-style zero initializer list anymore.

Reply:
No, it should not, if `S` defines a default ctor, `S obj = {}` should be calling that default ctor.

Reply:
However, if `S` is a C-style struct, then `S s = {}` will still fall back to c-style initailization logic according to this proposal because the ctor match will fail. In that case, we will still zero initialize.

Reply:
Can we also make a clear definition in this proposal that what is "C-style" struct?
I can see there is function in our code base, but it's more clear that people can find the definition in the doc as well.

Reply:
For the implementation purpose, we hope it's  very clear that which path is a special case that we could potentially remove in the future.